lib/checkother.cpp

```diff
@@ -485,6 +485,11 @@ void CheckOther::checkRedundantAssignment()
                 else
                     start = tok->findExpressionStartEndTokens().second->next();
 
+                // Is this a variable declaration with initialization?
+                const Token* defineAndInitTok = start->tokAt(-3);
+                if(Token::Match(defineAndInitTok, "; %var% =") && defineAndInitTok->linenr() == start->linenr())
```

Comment:
I am not sure we should bailout for all initializations. I suggest that you only use this early continue if rhs is `0`, `{0}`, `NULL` or `nullptr`.

Reply:
I think we have some bailout for `0`, `nullptr` and `NULL` already.. so maybe we can just extend that.

But I have the feeling that it might be nice to write warnings anyway here. The initialization _is_ redundant. We should use a different id and message though so people can suppress it if they want to.

Reply:
Hmm ok...our coding rules as an example say that you must initialize any variable even if it gets overwritten in the next statement. So the current rule produces a lot of "false positives" in our case.
So should I copy the current redundantAssigment check and just check for that case? Or what is the prefered way?

Reply:
Just one other issue that I encounter but I don't know how to proceed here:
void main(std::vector<unsigned int> buffer)
{
    int index = 0;
    for (unsigned int value : buffer)
    {
        this->m_buffer[index] = value;
    }
}
I would expect that cppcheck would report a redundantAssignment or redundantCopy warning because index doesn't change in that loop.

Reply:
> Hmm ok...our coding rules as an example say that you must initialize any variable even if it gets overwritten in the next statement.

Yes exactly.. some people have such rule.

If you are interested ... the reason I do not want to have such initialization is that it prevents static/dynamic analysis tools from detecting bugs in your code. I do not think it's good to hide bugs in your code.

> So should I copy the current redundantAssigment check and just check for that case? Or what is the prefered way?

I want that you set a flag. you can replace the "continue;" in your patch with "isInitialization = true" perhaps.

But are you sure you want to avoid the warnings for all initializations? How about:

    int *p = malloc(100);
    p = malloc(100);

> I would expect that cppcheck would report a redundantAssignment or redundantCopy warning because index doesn't change in that loop.

I agree. Can you create a Trac ticket? Or should I?

Reply:
Can you create a Trac ticket please? I don't have a Trac account yet.
As for the redundantAssigment check: when setting a flag where should this flag be checked? Bundle the check with a global setting or is there a special redundantAssignment check option the use can suppress such findings?

Reply:
> Can you create a Trac ticket please? I don't have a Trac account yet.

ok! I created ticket 9284

> As for the redundantAssigment check: when setting a flag where should this flag be checked?

We have this code in checkother.cpp:

                // warn
                if (hasCase)
                    redundantAssignmentInSwitchError(tok, nextAssign, tok->astOperand1()->expressionString());
                else
                    redundantAssignmentError(tok, nextAssign, tok->astOperand1()->expressionString(), inconclusive);

.. I suggest that you add one more function `redunantInitializationError`